platform/external-system-api/src/com/intellij/openapi/externalSystem/model/DataNode.java

```diff
@@ -279,9 +279,13 @@ private void readObject(ObjectInputStream in)
 
   @Override
   public int hashCode() {
-    int result = myChildren.hashCode();
-    result = 31 * result + myKey.hashCode();
+    int result = myKey.hashCode();
     result = 31 * result + getData().hashCode();
+    // We can't use myChildren.hashCode() because it's exponential.
```

Comment:
What do you mean by exponential? Do you mean that hashCode may be called many times for single object? For me it looks like the time is linear in size of DataNode (including all its children recursively), though it may be too much indeed.